### PR TITLE
Add program and runtime support for Durable Transaction Nonces

### LINKED ISSUE #1982: Blockhash in transactions makes cross chain atomic swaps unreliable
#### Problem

The blockhash in transactions makes cross-chain atomic swaps unreliable. The Solana half of the swap is valid for a short time. While the external chain transaction is replayable until it’s committed.  

#### Proposed Solution

Make the blockhash optional in transactions. These transactions would fully spend an account since they are replayable.

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

Using `recent_blockhash` values as transaction nonces imposes an unacceptably short time allowance for offline signing situations.

#### Summary of Changes

Implement a contract-base, compare-and-swap-like, mechanism for stashing `recent_blockhash` values. ala #6725.

Fixes #1982

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -891,6 +896,33 @@ impl Bank {
                 if lock_res.is_ok()
                     && !hash_queue.check_hash_age(&tx.message().recent_blockhash, max_age)
                 {
+                    // Too old! Let's see if we're using the Nonce program...
+                    let maybe_nonce_ix = &tx.message.instructions[0]; // Could search here... at least tx.msg.program_ids()...
+                    let program_id = tx.message.account_keys[maybe_nonce_ix.program_id_index as usize];
```
Comment (rob-solana) [inline runtime/src/bank.rs]: would prefer all these checks were done in the Nonce code, given a transaction...

File: sdk/src/nonce_state.rs [hunk 2]
```diff
@@ -0,0 +1,549 @@
+use serde_derive::{Deserialize, Serialize};
```
Comment (rob-solana) [inline sdk/src/nonce_state.rs]: the nonce program code belongs in runtime, like system_program?

File: sdk/src/sysvar/recent_blockhashes.rs [hunk 3]
```diff
@@ -74,6 +75,13 @@ where
     account
 }
 
+pub fn create_test_recent_blockhashes(start: usize) -> RecentBlockhashes {
```
Comment (rob-solana) [inline sdk/src/sysvar/recent_blockhashes.rs]: can you move this down in to `mod tests {` ?

File: sdk/src/nonce_state.rs [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+        if !signers.contains(self.unsigned_key()) {
+            return Err(InstructionError::MissingRequiredSignature);
+        }
+
+        match self.state()? {
+            NonceState::Uninitialized => {
+                if lamports > self.account.lamports {
+                    return Err(InstructionError::InsufficientFunds.into());
+                }
+                self.set_state(&NonceState::Uninitialized)?;
```
Comment (rob-solana) [inline sdk/src/nonce_state.rs]: no need to set this, is already that...

File: sdk/src/nonce_state.rs [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+            },
+            NonceState::Initialized(meta, ref hash) => {
+                if lamports == self.account.lamports {
+                    if recent_blockhashes.contains(hash) {
+                        return Err(NonceError::NotReady.into());
```
Comment (rob-solana) [inline sdk/src/nonce_state.rs]: NotExpired is a better description

File: sdk/src/nonce_state.rs [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+                    }
+                    self.set_state(&NonceState::Uninitialized)?;
+                } else if lamports + meta.rent_exempt_lamports > self.account.lamports {
+                    return Err(InstructionError::InsufficientFunds.into());
+                } else {
+                    self.set_nonce(recent_blockhashes, &meta)?;
```
Comment (rob-solana) [inline sdk/src/nonce_state.rs]: why Nonce() here?  can you add a comment for the "why"?

File: sdk/src/nonce_state.rs [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+    }
+
+    fn nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        rent: &Rent,
+    ) -> Result<(), InstructionError> {
+        let meta = match self.state()? {
+            NonceState::Initialized(meta, ref hash) => {
+                if recent_blockhashes.contains(hash) {
```
Comment (rob-solana) [inline sdk/src/nonce_state.rs]: this can be relaxed.  as long as the hash isn't the *most recent* blockhash, we can advance it

**Comment by rob-solana** [comment]
you're jamming on this

File: runtime/src/bank.rs (hunk 1 above)
Comment (t-nelson) [reply runtime/src/bank.rs]: Sure.  I'll move them in there once everything is working

File: sdk/src/nonce_state.rs (hunk 2 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: `system_program` and `system_instruction` both live in `sdk`.  The only thing remaining in `runtime` is the instruction parsing.  I could split it up like that

File: sdk/src/sysvar/recent_blockhashes.rs (hunk 3 above)
Comment (t-nelson) [reply sdk/src/sysvar/recent_blockhashes.rs]: I'll try and see if Rust can still find it from other packages

File: sdk/src/nonce_state.rs (hunk 4 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: Ha!  Refactor fail...

File: sdk/src/nonce_state.rs (hunk 5 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: Agreed

File: sdk/src/sysvar/recent_blockhashes.rs (hunk 3 above)
Comment (garious) [reply sdk/src/sysvar/recent_blockhashes.rs]: It will only be able to find them if you put it in `mod tests` without a `#[cfg(test)]`.  Traditionally we haven't done that, but could start.

File: sdk/src/nonce_state.rs (hunk 6 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: It doesn't.  I had in my head for awhile that everything that modifies the account should be advancing the nonce.  Proposal convo shook that out, but changes hadn't made it here yet.

File: sdk/src/sysvar/recent_blockhashes.rs (hunk 3 above)
Comment (t-nelson) [reply sdk/src/sysvar/recent_blockhashes.rs]: Ah! That's it.  I knew I'd fought this fight before.  Same deal for tagging the function with `#[cfg(test)]` ?

Comment (garious) [reply sdk/src/sysvar/recent_blockhashes.rs]: indeed

File: sdk/src/nonce_state.rs [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+    nonce_instruction::{NonceError},
+    pubkey::Pubkey,
+    sysvar::recent_blockhashes::RecentBlockhashes,
+    sysvar::rent::Rent,
+};
+use std::collections::HashSet;
+
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq, Clone, Copy)]
+pub struct Meta {
+    rent_exempt_lamports: u64,
```
Comment (jstarry) [inline sdk/src/nonce_state.rs]: Is this necessary? I think you could conditionally require the rent account for withdraw instructions where `lamports < account.lamports`

File: sdk/src/nonce_state.rs [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+            )
+        }
+    }
+
+    fn nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        rent: &Rent,
+    ) -> Result<(), InstructionError> {
+        let meta = match self.state()? {
```
Comment (jstarry) [inline sdk/src/nonce_state.rs]: Look like there's an attack vector here if you don't enforce the nonce account to be a signer. Anyone could change the nonce before the nonced transaction is processed.

File: sdk/src/nonce_instruction.rs [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+}
+
+impl std::error::Error for NonceError {}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub enum NonceInstruction {
+    /// `Nonce` consumes a stored nonce, replacing it with a successor
+    ///
+    /// Expects 2 Accounts:
+    ///     0 - Initialized NonceAccount to be spent
```
Comment (jstarry) [inline sdk/src/nonce_instruction.rs]: Looks like `Nonce` is also used to _initialize_ a `NonceAccount` so I think this needs to be updated

File: sdk/src/nonce_state.rs (hunk 9 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: Yep, nice catch.  I noticed it on friday.  Fix is local, will get it up today

File: sdk/src/nonce_state.rs [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+    ) -> Result<(), InstructionError>;
+}
+
+impl<'a> NonceAccount for KeyedAccount<'a> {
+    fn set_nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        meta: &Meta,
+    ) -> Result<(), InstructionError> {
+        if recent_blockhashes.is_empty() {
```
Comment (jstarry) [inline sdk/src/nonce_state.rs]: I think this check could get moved to the beginning of `fn nonce(..)` and then `set_nonce` can be passed a blockhash directly

File: sdk/src/nonce_state.rs (hunk 8 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: The `NonceAccount` itself needs to be rent exempt.  My understanding is that this is signaled by maintaining a minimum balance specified by the network at account creation.

@rob-solana any clarifying comments?

Comment (jstarry) [reply sdk/src/nonce_state.rs]: Yeah I'm on the same page. Just commenting whether you need to store `rent_exempt_lamports` in the account meta. You only use it when processing withdraw instructions. Instead of storing that meta, you could require the rent account for withdraw instructions which withdraw less than the total lamports

Comment (ParthDesai) [reply sdk/src/nonce_state.rs]: @t-nelson what you could do is make `create_nonce_account` instruction require to be passed `AccountMeta` of rent sysvar, and while processing said instruction, just verify that account has sufficient balance, if not just return an error.

You can look up how stake and vote account creation has handled rent for more detailed ref.

File: sdk/src/nonce_state.rs (hunk 11 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: I'm actually leaning towards dumping `set_nonce` altogether.  The code reuse that inspired it no longer exists.

File: sdk/src/nonce_state.rs (hunk 8 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: Thanks!  I actually stole the rent exempt handling from stake 🙃.  Will take another look, though.

File: sdk/src/nonce_state.rs [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+    nonce_program::id,
+    nonce_instruction::{NonceError},
+    pubkey::Pubkey,
+    sysvar::recent_blockhashes::RecentBlockhashes,
+    sysvar::rent::Rent,
+};
+use std::collections::HashSet;
+
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq, Clone, Copy)]
+pub struct Meta {
```
Comment (jackcmay) [inline sdk/src/nonce_state.rs]: Why does `meta` exist if it only contains one element?  Named type for clarity?

File: sdk/src/nonce_state.rs [hunk 13]
```diff
@@ -0,0 +1,545 @@
...
+
+        match self.state()? {
+            NonceState::Uninitialized => {
+                if lamports > self.account.lamports {
+                    return Err(InstructionError::InsufficientFunds.into());
+                }
+            },
+            NonceState::Initialized(meta, ref hash) => {
+                if lamports == self.account.lamports {
+                    if *hash == recent_blockhashes[0] {
```
Comment (jackcmay) [inline sdk/src/nonce_state.rs]: Wouldn't this be a way to cancel it?  Do we not want to nonce cancellation?

File: sdk/src/nonce_state.rs (hunk 12 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: Named type and old habits die hard 🙂

File: sdk/src/nonce_state.rs (hunk 13 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: We must ensure that the nonce is replaced with a new value to prevent replay.

In what context are you referring to cancellation here?

Comment (jackcmay) [reply sdk/src/nonce_state.rs]: Replay?  I guess I'm confused about why we allow withdrawing in the first place if we can't withdraw all the lamports are we just providing a mechanism because someone might deposit too many lamports during account creation?

I would also think we would want a way to cancel a nonce account and extract all the lamports once its no longer needed.  When a nonce account is canceled it just means that lamport balance goes to zero and any further reference to it will fail

File: sdk/src/nonce_instruction.rs [hunk 14]
```diff
@@ -0,0 +1,479 @@
...
+    ///     1 - A system account to which the lamports will be transferred
+    ///     2 - RecentBlockhashes sysvar carrying a subset of the bank's
+    ///         BlockhashQueue
+    ///
+    /// The `u64` parameter is the lamports to withdraw, which must leave the
+    /// account balance above the rent exempt reserve or at zero.
+    Withdraw(u64),
+}
+
+pub fn create_nonce_account(
```
Comment (jackcmay) [inline sdk/src/nonce_instruction.rs]: Maybe `create_nonce_account` should create a transaction that both creates the system account and initialized the nonce.  That way it would be atomic and we don't have the initial lamport balance separated from the min lamports required check during initialize

**Comment by jackcmay** [comment]
Do nonces expire?

File: sdk/src/nonce_state.rs [hunk 15]
```diff
@@ -0,0 +1,545 @@
...
+    }
+
+    fn nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        rent: &Rent,
+    ) -> Result<(), InstructionError> {
+        let meta = match self.state()? {
+            NonceState::Initialized(meta, ref hash) => {
+                if *hash == recent_blockhashes[0] {
```
Comment (jackcmay) [inline sdk/src/nonce_state.rs]: Why is this an error, is there a problem pinning the latest blockhash?

**Comment by t-nelson** [comment]
@jackcmay required reading, https://github.com/solana-labs/solana/pull/6725 😉

**Comment by t-nelson** [comment]
> Do nonces expire?

No. They are live until consumed by another `Nonce` instruction or the funds are fully withdrawn and the account deleted

File: sdk/src/nonce_state.rs (hunk 15 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: Same reason as in `withdraw`.  If the stored nonce value is not replaced with a new value, it will allow a transaction to be replayed.

File: sdk/src/nonce_instruction.rs (hunk 14 above)
Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: There was originally a separate `Initialize` instruction to drive the account state from `Uninitialized` to `Initialized`.  Some of the tests that differentiated `Initialize` from `Nonce` had to be moved into runtime to prevent fee theft.  What was left didn't really warrant a separate instruction

File: sdk/src/nonce_state.rs (hunk 13 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: If the stored nonce value does not advance, it would allow a transaction using this featured to be resubmitted to the network (replayed) and succeed again.

Thanks for clarifying "cancellation", it's what I've been calling "close" in my head.  The `Withdraw` stuff came in with Rent Exemption, which I don't fully grok the minutia on yet.  I pretty much monkey-see-monkey-do'd it out of staking to get things moving.  I agree that an instruction that just does the full withdraw and allows the account to be cleaned up seems more ergonomic, but I'm not sure of the constraints of closing a rent exempt account.

File: sdk/src/nonce_instruction.rs (hunk 14 above)
Comment (jackcmay) [reply sdk/src/nonce_instruction.rs]: From a user's perspective, it would be nice if the API wrapped these into one atomic operation (create an account and initialize the nonce

File: sdk/src/nonce_state.rs (hunk 15 above)
Comment (jackcmay) [reply sdk/src/nonce_state.rs]: I guess it doesn't really matter anyway since the bank won't even use the nonce since transaction's `recent_blockhash` isn't "old"

File: sdk/src/nonce_state.rs (hunk 13 above)
Comment (jackcmay) [reply sdk/src/nonce_state.rs]: I guess the question is also do we want to give the nonce owner the ability to invalidate a transaction they've signed.  Seems like they already have this ability inadvertently by signing a transaction and submitting it before the first.  The first would then have a non-matching nonce and be rejected.

File: runtime/src/bank.rs [hunk 16]
```diff
@@ -891,6 +896,33 @@ impl Bank {
...
+                    if nonce_program::check_id(&program_id) {
+                        if let Ok(nonce_instruction::NonceInstruction::Nonce) = limited_deserialize(&maybe_nonce_ix.data) {
+                            let accounts = self.rc.accounts.load_accounts(
```
Comment (jackcmay) [inline runtime/src/bank.rs]: I think `self.rc.accounts.load_slow()` (bad name) would be better and more direct here

File: sdk/src/nonce_instruction.rs (hunk 14 above)
Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: The `Initialize` instruction was issued atomically (WRT user API).  I can't think of any reason off the top of my head not to issue a `Nonce` similarly.  I'll review my convo with @rob-solana to be sure I'm not forgetting something and toss it in

Comment (jackcmay) [reply sdk/src/nonce_instruction.rs]: Was thinking something like this: https://github.com/solana-labs/solana/blob/c706f9b2cd194f44bffa78e64e84550a75211ab2/programs/vest/src/vest_instruction.rs#L87

Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: Yep!  It used to look like this:
https://github.com/solana-labs/solana/pull/6845/commits/33b9954e269f4e871f10a38444ca47573585b370#diff-37a2ca5217495a5ada65ad8207fcc32bL100-L110

Comment (rob-solana) [reply sdk/src/nonce_instruction.rs]: a separate Initialize instruction just seemed unnecessary.  Nonce does exactly one job, it always requires the Nonce account signature, and the actual instructions are hidden behind Vec<Instruction> library calls.

Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: @rob-solana, @jackcmay is asking if `create_nonce_account()` can return a `Nonce` instruction in addition to the `system_instruction::create_account`, as a convenience to SDK users.  I can't think of any reason not to

Comment (rob-solana) [reply sdk/src/nonce_instruction.rs]: yes, it can.

**Comment by jackcmay** [comment]
One thing to be aware of with these changes is that old blockhashes always result in an accont_db lock and this kind of DDoS is not thwarted by requiring account fees.

**Comment by garious** [comment]
Why is this PR still a draft PR?  And why does this draft PR have reviewers?

**Comment by t-nelson** [comment]
Draft because not rebased to HEAD, cargo fmt or clippy. Which I'm wrapping up now.

Reviews from the curious?

**Comment by garious** [comment]
@t-nelson, fyi, https://github.com/solana-labs/solana/blob/master/CONTRIBUTING.md#draft-pull-requests

**Comment by t-nelson** [comment]
The reviews we unsolicited, but very much appreciated!

**Comment by t-nelson** [comment]
TODO
- [ ] Move runtime check logic into nonce module
- [ ] Charge fees for nonce transactions
- [ ] Forward fees through transaction checks instead of `HashAgeKind`
- [ ] Investigate ways to avoid loading accounts twice to check nonce usage

File: runtime/src/bank.rs [hunk 17]
```diff
@@ -906,18 +921,47 @@ impl Bank {
...
-                if lock_res.is_ok()
-                    && !hash_queue.check_hash_age(&tx.message().recent_blockhash, max_age)
-                {
-                    error_counters.reserve_blockhash += 1;
-                    Err(TransactionError::BlockhashNotFound)
-                } else {
-                    lock_res
+                match lock_res {
+                    Ok(()) => {
+                        let message = tx.message();
```
Comment (garious) [inline runtime/src/bank.rs]: This code looks super fragile. You add about 4 levels of branching within a `map()` closure, and those ~2^4 code paths are tested with a single test.  How about breaking this up into small functions and testing all the edge cases with separate unit tests. Might need to add some helper functions to avoid copy-pasta for setup code.

Also, I'd be surprised to see all that array indexing is safe.

File: runtime/src/bank.rs [hunk 18]
```diff
@@ -1070,15 +1114,20 @@ impl Bank {
...
                 Err(e) => Err(e.clone()),
-                Ok((accounts, loaders, _rents)) => {
+                Ok((accounts, loaders, _rents, hak)) => {
                     signature_count += u64::from(tx.message().header.num_required_signatures);
-                    self.message_processor
-                        .process_message(tx.message(), loaders, accounts)
+                    let res =
+                        self.message_processor
+                            .process_message(tx.message(), loaders, accounts);
+                    match res {
```
Comment (garious) [inline runtime/src/bank.rs]: `res.map(|_| hak.clone())`

File: sdk/src/nonce_instruction.rs [hunk 19]
```diff
@@ -0,0 +1,430 @@
...
+        rent::{self, Rent},
+        Sysvar,
+    },
+};
+use log::*;
+use num_derive::{FromPrimitive, ToPrimitive};
+use serde_derive::{Deserialize, Serialize};
+use solana_logger;
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, FromPrimitive, ToPrimitive)]
```
Comment (garious) [inline sdk/src/nonce_instruction.rs]: Don't need `Serialize` or `Deserialize`

File: sdk/src/nonce_instruction.rs [hunk 20]
```diff
@@ -0,0 +1,430 @@
...
+pub enum NonceError {
```
Comment (garious) [inline sdk/src/nonce_instruction.rs]: Take `thiserror` for a spin?

File: sdk/src/nonce_state.rs [hunk 21]
```diff
@@ -0,0 +1,610 @@
+use crate::{
```
Comment (garious) [inline sdk/src/nonce_state.rs]: Does `nonce_state.rs` need to be in the SDK? Why can't `runtime` add `solana-nonce-program` as a direct dependency?

**Comment by garious** [comment]
Overall, this PR got way [too big](https://github.com/solana-labs/solana/blob/master/CONTRIBUTING.md#how-big-is-too-big). It's well into mega-PR territory.  Why not get the nonce program merged first, where there's no chance of getting into rebase hell. And then work on the runtime changes separately.

File: runtime/src/bank.rs (hunk 17 above)
Comment (t-nelson) [reply runtime/src/bank.rs]: Yeah. This isn't my favorite chunk.  I'll break up the test and see if I can't find some logical spots to split out this logic.  The indexing stuff should've been in the TODO.

File: sdk/src/nonce_instruction.rs (hunk 20 above)
Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: I'll take a peek at https://github.com/solana-labs/solana/pull/7177

File: sdk/src/nonce_state.rs (hunk 21 above)
Comment (t-nelson) [reply sdk/src/nonce_state.rs]: It doesn't.  The only thing linking`nonce_instruction.rs` back to `nonce_state.rs` is the state size convenience function, which I can duplicate.

Can you elaborate on "direct dependency"?   I tried to follow `system_program`

**Comment by t-nelson** [comment]
Thanks for reviewing!

Totally agreed this got out of hand. On the plus side, I'm much more comfortable in runtime now, so ought to be able to avoid a repeat going forward.

In my initial naivety I had a goal of keeping this entirely contained in the program.  The guaranteed fee theft crushed that

**Comment by jackcmay** [comment]
@t-nelson FYI: I have changes to #7118 queued up and waiting for this PR to land

**Comment by rob-solana** [comment]
> Thanks for reviewing!
> 
> Totally agreed this got out of hand. On the plus side, I'm much more comfortable in runtime now, so ought to be able to avoid a repeat going forward.
> 
> In my initial naivety I had a goal of keeping this entirely contained in the program. The guaranteed fee theft crushed that

what's the plan for this PR moving forward?  chop it up or try to land as-is?

**Comment by t-nelson** [comment]
I'm almost done addressing @garious's review. Could probably split fairly easily along the SDK/Runtime boarder after that if you like

**Comment by rob-solana** [comment]
> I'm almost done addressing @garious's review. Could probably split fairly easily along the SDK/Runtime boarder after that if you like

my goal in asking was to get a feel for whether Nonce's are intrusive enough into the runtime that we should just make 'em a feature of system, which would reduce the number of signatures required for transactions using the Nonce (Nonce could pay fees).

**Comment by t-nelson** [comment]
The runtime deltas are pretty much just the actual durable tx nonce usage check in the case a tx fails `check_hash_age` and a forwarding the result thereof to work around getting fees when a durable nonce was used.

I'll push what I have committed so far

**Comment by garious** [comment]
wow, that looks much nicer, thanks!

File: sdk/src/nonce_state.rs (hunk 12 above)
Comment (jackcmay) [reply sdk/src/nonce_state.rs]: Are you planning to remove `Meta`?

Comment (t-nelson) [reply sdk/src/nonce_state.rs]: Nope. https://github.com/solana-labs/solana/issues/7305

**Comment by t-nelson** [comment]
This is functionally ready now.

Cosmetically, I'd like to break up the largest of the bank tests a bit and get `nonce_state` out of SDK.  The latter is tricky due to the `NonceState`'s size needing exposed for the account creation instruction.  Going to `runtime` or `programs/nonce` will close a circular dependency.  A new `system_instruction::CreateNonceAccount` is the most sane thing I can think of, but seems ham fisted.  Any suggestions greatly appreciated!

I'm not sure either of the cosmetics are worth holding this up any further

**Comment by garious** [comment]
Code looks nice and maintainable, though consider more use of `Result::map_err` to remove some matches, and use the `?` operator on `Option`s to remove the `and_then` chaining.

File: core/src/banking_stage.rs [hunk 22]
```diff
@@ -443,15 +447,15 @@ impl BankingStage {
     fn record_transactions(
         bank_slot: Slot,
         txs: &[Transaction],
-        results: &[transaction::Result<()>],
+        results: &[TransactionProcessResult],
```
Comment (mvines) [inline core/src/banking_stage.rs]: The `TransactionProcessResult` plumbing in this PR is a great example of something that could be factored out and landed immediately.     It's cluttering up the meat of this PR and it barely even needs review

File: genesis-programs/src/lib.rs [hunk 23]
```diff
@@ -82,6 +84,7 @@ pub fn get_programs(operating_mode: OperatingMode, epoch: Epoch) -> Option<Vec<(
                 // a future hard fork.
                 Some(vec![
                     solana_config_program!(),
+                    solana_nonce_program(),
```
Comment (mvines) [inline genesis-programs/src/lib.rs]: We're gonna want this in epoch 0

File: runtime/src/accounts.rs [hunk 24]
```diff
@@ -822,13 +863,13 @@ mod tests {
         assert_eq!(error_counters.account_not_found, 0);
         assert_eq!(loaded_accounts.len(), 1);
         match &loaded_accounts[0] {
-            Ok((transaction_accounts, transaction_loaders, _transaction_rents)) => {
+            (Ok((transaction_accounts, transaction_loaders, _transaction_rents)), _hak) => {
```
Comment (mvines) [inline runtime/src/accounts.rs]: `_hak`?

Comment (mvines) [reply runtime/src/accounts.rs]: oh `_hash_age_kind`.  Just do that, don't make your reader guess

**Comment by mvines** [comment]
This PR has gotten too big to fail :-/

File: core/src/banking_stage.rs (hunk 22 above)
Comment (t-nelson) [reply core/src/banking_stage.rs]: Sure! I can do a bit of history cleanup

**Comment by garious** [comment]
What are we blocked on here?

**Comment by t-nelson** [comment]
I've addressed @mvines' comments and did a quick history rework to ease review.  Will push once this local test run finishes.

File: core/src/banking_stage.rs (hunk 22 above)
Comment (mvines) [reply core/src/banking_stage.rs]: For next time, please just bust this out immediately into a completely different PR and get it out of the way.   I really want to review this PR but at ~2k lines of new code I just can't do it.   20 smaller PRs that gradually introduce the feature, no problem!

Comment (garious) [reply core/src/banking_stage.rs]: I second. Considering how long this PR has been open, even 40 PRs wouldn't have been a surprise.

Comment (t-nelson) [reply core/src/banking_stage.rs]: Thirded!  Admittedly not my best showing and full apologies for that!  I'm not sure I could've got it up to 40 independent parts, but even the four commits that it is now would've been much better

Comment (mvines) [reply core/src/banking_stage.rs]: Forthed!

**Comment by t-nelson** [comment]
Any approvals incoming or is it merge time?

File: sdk/src/nonce_instruction.rs [hunk 25]
```diff
@@ -0,0 +1,413 @@
...
+    ///
+    /// Expects 3 Accounts:
+    ///     0 - A NonceAccount
+    ///     1 - A system account to which the lamports will be transferred
+    ///     2 - RecentBlockhashes sysvar carrying a subset of the bank's
+    ///         BlockhashQueue
+    ///
+    /// The `u64` parameter is the lamports to withdraw, which must leave the
+    /// account balance above the rent exempt reserve or at zero.
+    Withdraw(u64),
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: Ah, right!  I meant to revisit this as it was brought up earlier.  Agreed on the above, though `Meta` will likely be back with Authorized Noncer.

Also, we have this same pattern in the stake program.  I can open an issue for its removal if you like

[the remark this answers is not in the document]
Comment (mvines) [reply sdk/src/nonce_instruction.rs]: Oh yep it comes back with the Noncer.  Empty struct for now?

[the remark this answers is not in the document]
Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: That's the plan!

[the remark this answers is not in the document]
Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: cc/ @jackcmay RE #7329

**Comment by t-nelson** [comment]
24hr merge ⏲ starts now

**Comment by mvines** [comment]
But it's so green.   👿 loves to press green buttons.

**Comment by garious** [comment]
:shipit:

**Comment by t-nelson** [comment]
The audience has spoken!
